### PR TITLE
fix(lifecycle): reconcile saved OpenShell and container state

### LINKED ISSUE #11646: [Bug] Reconcile saved sandbox phase with runtime state during Portable recovery
Parent epic: https://github.com/NVIDIA/NemoClaw/issues/11573

An inconsistent saved phase can prevent recovery or report a successful start while the container remains stopped.

At NemoClaw `e9cf24220532e33e704298632d9006b30254efdd`, isolated OpenShell 0.0.116 tests reproduced three failures:

- A real image configuration failure reached Error. A later start failed because the saved phase was not Stopped.
- A healthy container with an injected saved Error refused both start and exec.
- An exited container with an injected saved Ready reported successful start but remained exited. Exec exceeded the harness deadline.

Restoring the correct Stopped phase allowed start and exec. Separate NemoClaw tests show that an unhealthy workload in a running container receives health polling without workload relaunch.

These controls use a minimal Ubuntu workload and a disposable database. They establish the state mismatch behavior, not the cause of John's failure or Hermes recovery performance.

## Reproduction

Run the disposable OpenShell healthy control described below. Repeat with deliberately mismatched saved phases in that disposable runtime. Separately use an image without the required OCI user to reach Error naturally. Retain container state, saved phase, command output, and exit status for each run.

## Acceptance

- Reconcile saved phase, container state, and managed workload readiness before reporting success.
- Recover through the owning lifecycle API when recovery is supported. Preserve actionable failure evidence when it is not.
- Do not reset database phase or directly mutate an unqualified container as a repair.
- Preserve timeout rollback, identity checks, and retained user data.
- Repeat healthy stop/start, Error, Ready/exited, and unhealthy-workload controls against the fix.

## Recorded evidence

OpenShell 0.0.116: saved Error + running container refused start/exec; saved Ready + exited container returned start success while the container stayed exited. A correct Stopped control restarted and executed successfully.

Reproduction baseline: `e9cf24220532e33e704298632d9006b30254efdd`. Canonical main was checked at `cad153cc1b761353e4a9590da7fa2f5121913056` before filing. The listed behavior owners are unchanged between those revisions; after-fix validation is pending.

- [Behavior owner: `src/lib/onboard/experimental/hermes-portable-lifecycle.ts`](https://github.com/NVIDIA/NemoClaw/blob/e9cf24220532e33e704298632d9006b30254efdd/src/lib/onboard/experimental/hermes-portable-lifecycle.ts)

## Related work

Related historical reports #11248, #11302, and #8112 are closed. This issue tracks the reproduced saved-phase/runtime mismatch at the stated newer baseline; it does not reopen their distinct manifest migration or stop-settlement scopes.

## Delivery

One focused PR will link this issue with `Fixes`. It must include the before/after reproduction, relevant regression evidence, and review and CI outcomes for its final commit. Preserve installed kiosk launchers and existing user sandboxes.

### PR DESCRIPTION
**Pull request by prekshivyas**
## Outcome

Hermes Portable recovery reconciles a saved `Stopped` phase with an already-running receipt-owned container through OpenShell start. Saved `Error`, or saved `Ready` with a container that is not running, fails before mutation. Recovery preserves pre-existing running containers when reconciliation or health checks fail.

## Reason

Recovery previously used only the container running flag. It skipped supported saved-phase reconciliation and attempted unsupported recovery. OpenShell 0.0.116 can report start success for saved `Ready` while the container remains exited.

### Related issues

Fixes #11646

Part of #11573.

## Changes

- Classify the saved OpenShell phase together with qualified container state, and use OpenShell for lifecycle mutations.
- Refresh the phase after start, reusing full qualification when available. Retain identity, authenticated-health, and rollback checks.
- Reject unsupported mismatches with diagnostics, without resetting OpenShell storage or directly starting Podman.
- Cover supported reconciliation, rejected states, failed starts, and preservation of existing workloads in the owning lifecycle tests. Keep their receipt builder in the existing test companion within the unchanged file-size budget.
- Document the recovery and rollback contract. Resolve the merge conflict with current `main` while retaining the corrected rollback fixture and upstream lint cleanup.

## Verification

Candidate: `737c00e82620eb05e5e21b34e8d8ecdc30f44348`. Base: `475a20e00db5984d026b70d0a01654a2822b6a7a`.

- Five state regressions and one corrected-fixture rollback regression failed before the original production fix.
- All 47 lifecycle tests pass. The post-start observation regression failed before the duplicate-read repair and passes afterward. Deliberate mutations independently verified direct-Podman rejection, preservation of pre-existing containers after health failure, and rejection of `Ready`/`stopping`; all temporary mutations were removed.
- `npm run docs`: zero errors and five existing warnings. Normal signed commit hooks, pre-push publication validation, and CLI TypeScript checks passed. Validator inputs match canonical `main`; installed dependencies match the independent reference byte for byte. All six commits are GitHub Verified.
- Isolated OpenShell 0.0.116 controls confirmed supported `Stopped`/running reconciliation, false start success for `Ready`/exited, and refusal of stop/start for `Error` with running or exited containers. Healthy control operations passed, and disposable cleanup was verified. These controls make no installed-seat timing claim.
- [Portable rootless E2E](https://github.com/NVIDIA/NemoClaw/actions/runs/34777204907) passed on this candidate. Its receipt verifies authenticated recovery, restored forwarding, final stop settlement, and all 13 phases including cleanup.
- [Required CI](https://github.com/NVIDIA/NemoClaw/actions/runs/34777204823) and [managed-image validation](https://github.com/NVIDIA/NemoClaw/actions/runs/34777204854) passed on this candidate. CI includes all 12 CLI shards and coverage.
- [Onboarding repair/resume E2E](https://github.com/NVIDIA/NemoClaw/actions/runs/34777271245/attempts/2) passed all three OpenClaw/Hermes executions, with exact candidate/attempt receipts and every phase and cleanup successful. It used successful immutable image contracts from direct parent `0557053`; only documentation and tests differ, with production code, image inputs, and dependencies unchanged. Current-head image validation also passed. The first attempt missed the ordinary Docker Hermes gateway readiness deadline; that failure and its successful cleanup remain retained with an unclassified underlying cause. One bounded rerun passed without code, timeout, or test changes.

## Review notes

Self-review covers the sensitive lifecycle source and tests, phase classification, OpenShell mutation ownership, preservation of existing workloads, rollback authority, final qualification, and the command reference.

[CodeRabbit's full review](https://github.com/NVIDIA/NemoClaw/pull/11668#issuecomment-5655503040) completed for this candidate with no actionable issues. The valid ownership-test finding is fixed. The independently reproduced inherited rollback finding was withdrawn, and both existing review threads remain resolved.

All nine findings files, summaries, and E2E recommendations from [Advisor 34778003881](https://github.com/NVIDIA/NemoClaw/actions/runs/34778003881) were read and verified against this candidate and base. Eight specialists are clear. The documentation specialist proposed replacing “saved OpenShell phase” with “current OpenShell phase”; this was assessed as a terminology false positive. OpenShell’s persisted phase is the state returned by its list/get API, and the accepted issue’s disposable-database controls prove it can disagree with the live container. The reference names OpenShell as its owner and directs resolution through OpenShell; it does not describe the phase as a Portable receipt field. The existing wording is accurate, so no change is required.

Both recommended E2E jobs, `onboard-repair` and `onboard-resume`, are satisfied by the current-candidate evidence above. All valid in-scope findings are resolved. Historical failed or incomplete attempts remain preserved, including the parent candidate’s Advisor submission-protocol failures after its one full rerun. No test, coverage, or review gate was waived.

The PR remains a draft. No merge or approval is claimed.

---
Signed-off-by: Prekshi Vyas <34834085+prekshivyas@users.noreply.github.com>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved portable recovery for stopped containers and restart requirements.
  - Added validation for inconsistent lifecycle states before recovery begins.
  - Improved handling of failed starts and post-start status updates.
  - Corrected results for already-running environments.
  - Expanded recovery coverage for stopped, unsupported, and failed states, including rollback scenarios.
  - Preserved running workloads and prevented recovery when saved and actual lifecycle states conflict.

- **Documentation**
  - Clarified phase reconciliation, startup handling, rollback behavior, and authenticated health checks during portable recovery.

### CODE REVIEW

**Comment by prekshivyas** [comment]
@coderabbitai full review

Please review `023aed9078c3d7f0c81a00ce1219d1a424ed590f`: reconcile saved Stopped through OpenShell, fail unsupported Error/Ready-exited states before mutation, and preserve pre-existing running containers during failure. Real isolated runtime controls and all 45 lifecycle tests pass as described in the PR.

**Comment by prekshivyas** [comment]
Review follow-up for `023aed9078c3d7f0c81a00ce1219d1a424ed590f`: this PR remains a draft and is incomplete.

- The original defect evidence records six failures before the fix and 45 passing lifecycle tests afterward. The isolated OpenShell controls remain unchanged.
- CodeRabbit completed its full review with no actionable findings. Its docstring percentage warning is advisory.
- [CI attempt 2](https://github.com/NVIDIA/NemoClaw/actions/runs/34747960863/attempts/2) passed the previously failing shard, merged tests, and coverage ratchet (83.5% lines). [Coverage publication failed with HTTP 500](https://github.com/NVIDIA/NemoClaw/actions/runs/34747960863/job/103702512146). The required check remains failed; no validation settings were relaxed.
- I read all available findings, summaries, and E2E recommendations from [Advisor attempts 1 and 2](https://github.com/NVIDIA/NemoClaw/actions/runs/34749019544). This covers all nine specialists across the two attempts, but neither attempt is complete. Attempt 1 omitted documentation after a submission-protocol failure. The authorized full rerun omitted migration after another submission-protocol failure: job `103702513681`, expected artifact `pr-review-specialist-migration-completion-2`. A successful migration review from attempt 1 does not clear the missing artifact from attempt 2. No further Advisor rerun was requested.
- Valid findings are grouped into two pending repairs: remove the unused `startupArgv` wrapper/import, and update the Hermes `recover` reference for saved `Stopped` reconciliation and unsupported mismatches. The two documentation findings share one cause. Both repairs are now applied in the assigned local worktree and remain uncommitted and unpublished. The revised worktree passes all 45 lifecycle tests, focused lint and formatting checks, and `npm run docs` (zero errors; five warnings). Generated-output comparison confirms that only the Hermes command reference changes; OpenClaw and Deep Agents outputs are byte-identical. Publication remains blocked while Advisor collection is incomplete.
- The first-attempt timing finding does not require a behavior change: `containerAction=reused` describes the already-running container, while `containerStartCount=1` records the OpenShell request that reconciles its saved phase. The sibling Portable timing owner uses the same container-state meaning. The first-attempt request to inline the receipt builder would grow the test from 1,464 to at least 1,534 lines, beyond the enforced 1,500-line limit. The existing companion fixture is retained. Both concerns were absent from the second-attempt reviews.
- The recommended `onboard-repair,onboard-resume` E2E dispatch returned HTTP 500. Repeated readback found no matching run for correlation `c641b41b-aaf6-4048-9d4a-28c80f42880e`; this evidence is incomplete. These standard-profile targets supplement the direct Hermes Portable regression evidence.

The published candidate is unchanged. CI publication, complete Advisor evidence, publication of the locally validated repair batch, and the recommended E2E evidence remain outstanding. No merge or readiness change was performed.

**Comment by prekshivyas** [comment]
@coderabbitai full review

Please review `1eff55a6b8d541f208075b5ef7a6d31514a93b0a`. This candidate includes the saved-phase recovery fix, the corrected Portable Hermes recovery reference, and removal of the unused test wrapper. All 45 lifecycle tests, the docs build, normal signed-commit hooks, pre-push publication validation, and CLI typecheck passed. The canonical validation dependency from merged #11660 is included unchanged.

**Comment by prekshivyas** [comment]
All nine specialist findings, summaries, and E2E recommendations from [Advisor run 34761276989](https://github.com/NVIDIA/NemoClaw/actions/runs/34761276989) have been read. That run reviewed `023aed9078c3d7f0c81a00ce1219d1a424ed590f`; the published repair is now `1eff55a6b8d541f208075b5ef7a6d31514a93b0a` and needs its own CI/reviews.

- Documentation finding `F-documentation-standard-work-f78ad31e44f85a8e2a00` is valid and fixed by the published Hermes recovery reference update. The earlier unused-wrapper finding is fixed too. All 45 lifecycle tests, docs build, normal commit hooks, pre-push publication validation, and CLI typecheck passed.
- Migration finding `F-migration-completion-f153384ec57733fed18b` proposes using stop to turn saved `Error` into `Stopped`. New real OpenShell 0.0.116 controls show that this transition is unavailable: for both a running and an exited receipt-equivalent disposable container, `sandbox stop` exits 1 with `sandbox must be Ready to stop (current phase: Error)`. Start also exits 1 because it requires `Stopped`; exec refuses `Error`. Healthy initial exec and final Stopped/start/exec controls passed. Fault injection touched only the disposable database. The sandbox and network were removed and their absence verified. The accepted issue explicitly requires actionable refusal when the owning API cannot recover the state, so this finding does not justify another recovery mutation.
- Operability finding `F-operability-recovery-576314a60b03aa75168f` conflates the physical container action with the OpenShell request. `containerAction=reused` describes the already-running container. The existing `containerStartCount=1` and `recovered` result distinguish saved-phase reconciliation, and the caller emits these fields. A new timing action would change the established container-state meaning without adding missing evidence.
- Reduction finding `F-reduction-simplification-1a3d9d2cdf8fc6b193dd` repeats the request to inline the receipt builder. The measured inline form exceeds the enforced 1,500-line test-file limit. The existing owner-specific fixture companion is permitted by `test/README.md` and preserves one receipt-construction owner; adding a second consumer just to justify it would add unrelated work.

The failed-CI retry for the preceding candidate passed. [Current candidate CI](https://github.com/NVIDIA/NemoClaw/actions/runs/34761618562) and [recommended onboard-repair/onboard-resume E2E](https://github.com/NVIDIA/NemoClaw/actions/runs/34761772523) are now running or queued. The PR remains draft; missing or unfinished evidence is not treated as clear.

File: src/lib/onboard/experimental/hermes-portable-lifecycle.test.ts [hunk 1]
```diff
@@ -355,6 +277,57 @@ afterEach(() => {
...
+      expect(() =>
+        withMcpLifecycleLockSync(
+          SANDBOX,
+          () => recoverHermesPortableSandboxLifecycle(SANDBOX, lifecycleContext(), fixture.deps),
+          { stateDir: path.join(stateDir, "state") },
+        ),
+      ).toThrow(`saved OpenShell phase ${initialPhase}`);
+      expect(openshellMutationCalls(fixture.captureOpenShell, "start")).toHaveLength(0);
+      expect(openshellMutationCalls(fixture.captureOpenShell, "stop")).toHaveLength(0);
+      expect(fixture.launchOpenShell).not.toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (prekshivyas) [reply src/lib/onboard/experimental/hermes-portable-lifecycle.test.ts:299]: Fixed in `f1a7f7e8104d6813964b1e04fb824784de1688ef`. The three tests (five saved-phase cases) now explicitly reject direct Podman `start` and `stop` calls while allowing inspection. A deliberate direct-Podman-start mutation passed all five cases before these assertions and fails all five afterward; the mutation was removed. All 45 lifecycle tests, focused lint/format, and normal commit hooks pass.

**Comment by prekshivyas** [comment]
@coderabbitai full review

Please review the current candidate `f1a7f7e8104d6813964b1e04fb824784de1688ef`, including the nine-line repair for discussion_r3999847477. The saved-phase cases now assert no direct Podman start/stop. Mutation testing demonstrates that the old tests missed direct Podman start and the new assertions reject it; normal lifecycle tests pass 45/45. Production and documentation are unchanged from `1eff55a`, whose required CI, managed-image validation, Portable rootless E2E and all nine Advisor specialist reviews completed successfully. New candidate CI and Advisor evidence remain required; the PR stays draft.

File: src/lib/onboard/experimental/hermes-portable-lifecycle.ts [hunk 2]
```diff
@@ -1544,7 +1555,7 @@ export function recoverHermesPortableSandboxLifecycle(
             )
           );
         }
-        startedByRecovery = true;
+        startedByRecovery = !wasRunning;
```
[the remark this answers is not in the document]
Comment (prekshivyas) [reply src/lib/onboard/experimental/hermes-portable-lifecycle.ts:1558]: I reproduced this exact zero-effect `start` response against both canonical base `b83c78e6693034967c16a5cd45f576b539b929d8` and candidate `f1a7f7e8104d6813964b1e04fb824784de1688ef`, using the same focused fixture and swapping only the production lifecycle module. The results are identical:

- Historical active receipt: one OpenShell start, no stop-assist attempt, no OpenShell stop, no Podman lifecycle mutation, and `OpenShell start did not start the receipt-owned container`.
- Successor receipt with retained transaction authority: one OpenShell start, one refused stop-assist attempt, no OpenShell stop, no Podman lifecycle mutation, and an aggregate preserving both the primary `post-start-authority` failure and the stop-assist failure. The exec stub follows OpenShell's real `Stopped`-phase refusal.

The base already sets `startedByRecovery = true` before the same retained-currentness check. This PR changes that assignment to `!wasRunning`, which preserves identical behavior for the reported initially-exited case and prevents rollback of a pre-existing running container. The added phase observation occurs after the check identified here. Existing tests also require bounded rollback when a recovery start takes effect but its first inspection fails.

I am classifying this as an inherited rollback edge case rather than a regression introduced by this saved-phase fix, in accordance with the repository follow-up contract. The primary diagnostic is retained, and the reproduced invalid-state case issued no stop. The known real OpenShell `Ready`/exited false-success case is rejected before start by this PR. No production change is made for this comment; the temporary diagnosis tests and base module replacement were restored. Please reclassify the finding on that basis.

**Comment by prekshivyas** [comment]
@coderabbitai full review

Resolved the conflict with main in f362217f865b0f6964e210896140bf5a9d0ddea0. The merge preserves the corrected saved-phase transition and upstream unused-variable cleanup; it also removes three fixture imports rejected by the new main lint rule. All 159 focused recovery, rollback, and forwarding tests passed, and normal signed-commit hooks passed. Fresh candidate CI and Advisor evidence are pending.

**Comment by prekshivyas** [comment]
@coderabbitai full review

Commit `0557053322c63a5e039b265102466ee24ec13974` resolves the valid findings from [Advisor 34774341488](https://github.com/NVIDIA/NemoClaw/actions/runs/34774341488). All nine findings files, complete summaries, and E2E recommendations were read.

- The duplicate post-start identity read is removed for receipts without retained transaction authority: full qualification already returns the current OpenShell phase. Retained transaction authority still receives its required explicit phase refresh. The strengthened existing regression failed before this repair (two post-start observations instead of one), then passed. All 160 focused tests pass, including 46 lifecycle cases.
- Pre-existing running containers remain outside recovery rollback ownership after saved `Stopped` reconciliation. The health-failure case now covers both initial `Ready` and `Stopped`. A deliberate rollback mutation stops the pre-existing container and fails the new case; the mutation was removed. The command reference now explicitly limits rollback to a previously stopped container that recovery started. Applying the suggested stop to an existing workload would violate that ownership boundary.
- The receipt builder stays in the existing lifecycle test companion. `test/README.md:25` explicitly permits one-test companions. Restoring the original local builder produces 1,548 lines, exceeding the unchanged 1,500-line limit in `ci/test-file-size-budget.json`; this file has no legacy exception. The helper comment records its consumer and placement reason. There is one receipt construction owner and no production consumer.
- The suggested `stop` then `recover` recipe is unsupported for saved `Error`. Retained isolated OpenShell 0.0.116 controls show `stop` exits 1 because the phase must be `Ready`, while `start` requires `Stopped` and `exec` also refuses `Error`, for both running and exited containers. Healthy stop/start/exec controls passed and disposable cleanup was verified. In the unchanged production stop owner, an exited `Error` is accepted as already stopped without changing that phase; a following recovery still correctly refuses it. No unsupported recovery instruction was added.

The previous candidate `f362217` passed [required CI](https://github.com/NVIDIA/NemoClaw/actions/runs/34773528670), [image validation](https://github.com/NVIDIA/NemoClaw/actions/runs/34773528682), [Portable rootless E2E](https://github.com/NVIDIA/NemoClaw/actions/runs/34773528649), and [onboarding repair/resume E2E](https://github.com/NVIDIA/NemoClaw/actions/runs/34774470719). The latter verified all three OpenClaw/Hermes executions and cleanup. Fresh CI and Advisor evidence are required for this repair commit; prior results are preserved as historical evidence.

**Comment by prekshivyas** [comment]
@coderabbitai full review

Commit `737c00e82620eb05e5e21b34e8d8ecdc30f44348` resolves documentation finding `F-documentation-standard-work-467ac6328ced1eb36864` from [Advisor 34776228940](https://github.com/NVIDIA/NemoClaw/actions/runs/34776228940). The command reference now says saved `Ready` with a container that is not running fails before mutation, matching the existing production guard. The existing rejection matrix now includes `Ready`/`stopping`.

All 47 lifecycle cases pass. A deliberate mutation narrowed the guard to `exited`; the new `stopping` case failed while the other three rejection cases passed. The mutation was removed. The documentation build passed with zero errors. This commit changes only documentation and tests; production code, image inputs, and dependencies are byte-identical to parent `0557053`.

The parent passed [required CI](https://github.com/NVIDIA/NemoClaw/actions/runs/34775437310), [Portable rootless E2E](https://github.com/NVIDIA/NemoClaw/actions/runs/34775437194), and [managed-image validation](https://github.com/NVIDIA/NemoClaw/actions/runs/34775437324/attempts/2). Image validation passed after one failed-job rerun; the original MCP discovery `offline: fetch failed` evidence and successful cleanup remain retained.

Advisor evidence for the parent remains incomplete. Attempt 1 produced eight specialist sets but omitted operability's required E2E result. The authorized full rerun produced seven sets, with verification and migration failing the submission protocol. All 45 available findings, summary, and E2E files across those attempts were read. The valid documentation finding is repaired even though the second documentation review was clear. Missing artifacts are not treated as clear; no further unchanged-candidate rerun was made. The user's bounded-retry policy explicitly authorizes publishing this independently validated repair and evaluating the new commit.

The previously resolved ownership-test finding and withdrawn inherited rollback finding retain their recorded dispositions. The PR remains a draft, and current-commit CI, Advisor, and onboarding repair/resume evidence are being collected.

**Review by jyaunches — APPROVED** [review]
